src/mesh/replicated_mesh.C

```diff
@@ -1417,5 +1417,151 @@ dof_id_type ReplicatedMesh::n_active_elem () const
+    {
+      // pop up an element
+      Elem * elem = list.back(); list.pop_back(); ++visited;
+
+      for (auto s : elem->side_index_range())
+      {
+        Elem * neighbor = elem->neighbor_ptr(s);
+        if (neighbor != nullptr && subdomains[neighbor->id()] == Elem::invalid_subdomain_id)
+        {
+          list.push_back(neighbor);
+          subdomains[neighbor->id()] = subdomain_counter;
+        }
+      }
+    }
+    std::cout << std::endl;
+    subdomain_counter++;
+  }
+  while (visited != n_elem());
+  // now subdomain_counter stores the number of disconnected subdomains
+
+  std::vector<std::vector<std::vector<Point>>> boundary_points(subdomain_counter);
+
+  // following boundary visiting algorithm only works with 2D
+  if (mesh_dimension() != 2)
+    libmesh_error_msg("Error: get_boundary_points only works for 2D now");
+
+  // get all boundary sides
+  // use a comparison functor to avoid run-time randomness due to pointers
+  struct boundary_side_compare
+  {
+    bool operator()(const std::pair<const Elem *, unsigned int> & lhs,
+                    const std::pair<const Elem *, unsigned int> & rhs) const
+    {
+      if (lhs.first->id() < rhs.first->id())
+        return true;
+      else if (lhs.first->id() == rhs.first->id())
+      {
+        if (lhs.second < rhs.second)
+          return true;
+      }
+      return false;
+    }
+  };
+  std::set<std::pair<const Elem *, unsigned int>, boundary_side_compare> boundary_elements;
+  for (const auto & elem : active_element_ptr_range())
+    for (auto s : elem->side_index_range())
+      if (elem->neighbor_ptr(s) == nullptr)
+        boundary_elements.insert(std::pair<const Elem *, unsigned int>(elem, s));
+
+  while (!boundary_elements.empty())
+  {
+    // get the first entry as the seed
+    const Elem * eseed = boundary_elements.begin()->first;
+    unsigned int sseed = boundary_elements.begin()->second;
+
+    // get the subdomain ID that these boundary sides attached to
+    subdomain_id_type subdomain_id = subdomains[eseed->id()];
+
+    // start visiting the mesh to find all boundary node with the seed
+    std::vector<Point> bpoints;
+    const Elem * elem = eseed;
+    unsigned int s = sseed;
+    std::vector<unsigned int> local_side_nodes = elem->nodes_on_side(s);
+    while (true)
+    {
+      boundary_elements.erase(std::pair<const Elem *, unsigned int>(elem, s));
+
+      // push all nodes on the side except the last one
+      for (unsigned int i = 0; i < local_side_nodes.size() - 1; ++i)
+        bpoints.push_back(*dynamic_cast<const Point *>(elem->node_ptr(local_side_nodes[i])));
+
+      // use the last node to find next element and side
+      const Node * node = elem->node_ptr(local_side_nodes[local_side_nodes.size() - 1]);
+      std::set<const Elem *> neighbors;
+      elem->find_point_neighbors(*node, neighbors);
+
+      // if only one neighbor is found (itself), this node is a cornor node on boundary
+      if (neighbors.size() != 1)
```

Comment:
Should this be ==, not !=?

Reply:
Oh, no, I get it.  If we *do* have more neighbors then one of them is what we're looking for and we can skip ourselves, if we *don't* then we're turning a corner on this element and we need to examine ourselves.